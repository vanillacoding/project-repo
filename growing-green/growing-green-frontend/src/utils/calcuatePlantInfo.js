import differenceInDays from 'date-fns/differenceInDays';
import addHours from 'date-fns/addHours';

export function calculatePlantInfo(plant, weather = 'Clear', currentDate) {
  const {
    createdAt,
    updatedAt,
    waterGuage,
    isBlindUp,
    isSunPlant,
    sunGuage,
    penaltyPoints,
    growthStage,
  } = plant;

  const updates = {
    sunGuage: sunGuage.currentGuage,
    waterGuage: waterGuage.currentGuage,
    penaltyPoints: penaltyPoints,
    growthStage,
  };

  const createdDate = new Date(createdAt);
  const lastLoginDate = new Date(updatedAt);
  const today = new Date(currentDate);

  const watering = waterGuage.defaultGuage;
  const NumberOfElapsedDate = differenceInDays(today, createdDate);

  const passedDays = [];
  const wateringDays = [];

  for (let i = 0; i < NumberOfElapsedDate; i++) {
    const passedDay = addHours(createdDate, 24 * (i + 1));

    passedDays.push(passedDay);
  }

  updates.growthStage = calculateGrowth(passedDays.length, growthStage);

  for (let i = 0; i < NumberOfElapsedDate / watering; i++) {
    const wateringDay = addHours(createdDate, 24 * ((i + 1) * watering));

    wateringDays.push(wateringDay);
  }

  const passDaysAfterUpdate = passedDays.filter((day) => day > lastLoginDate);
  const wateringDaysAfterUpdate = wateringDays.filter(
    (day) => day > lastLoginDate && day < today,
  );

  const wateringCount = wateringDaysAfterUpdate.length;

  if (wateringCount >= 1) {
    updates.waterGuage = 0;
    updates.penaltyPoints += wateringCount;
  }

  if (weather === 'rainy') {
    updates.waterGuage = waterGuage.defaultGuage;
  }

  if (isSunPlant === true) {
    if (isBlindUp === true) {
      const increasedGuage = sunGuage.currentGuage + passDaysAfterUpdate.length;
      const defaultGauge = sunGuage.defaultGuage;

      updates.sunGuage =
        increasedGuage > defaultGauge ? defaultGauge : increasedGuage;
    }

    if (isBlindUp === false) {
      const decreasedGuage = sunGuage.currentGuage - passDaysAfterUpdate.length;

      updates.sunGuage = decreasedGuage > 0 ? decreasedGuage : 0;

      if (decreasedGuage < 0) {
        updates.penaltyPoints = penaltyPoints - decreasedGuage;
      }
    }
  }

  if (isSunPlant === false) {
    if (isBlindUp === true) {
      const increasedGuage = sunGuage.currentGuage + passDaysAfterUpdate.length;
      const defaultGauge = sunGuage.defaultGuage;

      updates.sunGuage =
        increasedGuage > defaultGauge ? defaultGauge : increasedGuage;

      if (increasedGuage > defaultGauge) {
        const decreasePoints = increasedGuage - defaultGauge;
        updates.penaltyPoints = penaltyPoints - decreasePoints;
      }
    }

    if (isBlindUp === false) {
      const decreasedGuage = sunGuage.currentGuage - passDaysAfterUpdate.length;

      updates.sunGuage = decreasedGuage > 0 ? decreasedGuage : 0;
    }
  }

  const updatedPlant = {
    ...plant,
    sunGuage: {
      ...plant.sunGuage,
      currentGuage: updates.sunGuage,
    },
    waterGuage: {
      ...plant.waterGuage,
      currentGuage: updates.waterGuage,
    },
    penaltyPoints: updates.penaltyPoints,
    growthStage: updates.growthStage,
  };

  return updatedPlant;
}

function calculateGrowth(passedDays, current) {
  const ADULT_STAGE = 3;

  if (passedDays >= 14) {
    return ADULT_STAGE;
  } else if (passedDays >= 7) {
    return current + 1 >= 3 ? 3 : current + 1;
  } else {
    return current;
  }
}
