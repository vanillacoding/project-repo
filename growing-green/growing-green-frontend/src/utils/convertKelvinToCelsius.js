export function convertKelvinToCelsius(kelvin) {
  return Math.floor(kelvin - 273.15);
}
