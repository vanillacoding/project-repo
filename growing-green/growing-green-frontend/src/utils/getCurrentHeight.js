export const getCurrentHeight = (temp, height) => {
  const borderWidth = 7;
  const totalTemperature = 70;
  const minusTemperature = 30;

  return (
    ((Number(temp) + minusTemperature) / totalTemperature) * height +
    borderWidth
  );
};
