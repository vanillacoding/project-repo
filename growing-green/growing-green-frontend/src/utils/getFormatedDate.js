import { format } from 'date-fns';

export const getFormatedDate = (date) => {
  const dateString = format(date, 'yyyy MMM EEEE dd');
  const dateList = dateString.split(' ');

  return {
    year: dateList[0],
    month: dateList[1],
    week: dateList[2],
    day: dateList[3],
  };
};
