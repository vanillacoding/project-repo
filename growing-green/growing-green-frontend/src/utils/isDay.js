export function isDay(time) {
  if (time >= 6 && time < 18) {
    return true;
  }

  return false;
}
