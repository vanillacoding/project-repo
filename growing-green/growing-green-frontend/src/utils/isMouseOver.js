export function isMouseXOver(mouseX, startX, EndX) {
  if (mouseX > startX && mouseX < EndX) {
    return true;
  }

  return false;
}

export function isMouseYOver(mouseY, startY, endY) {
  if (mouseY > startY && mouseY < endY) {
    return true;
  }

  return false;
}
