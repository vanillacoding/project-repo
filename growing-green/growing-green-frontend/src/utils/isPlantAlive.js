export function isPlantAlive(point) {
  if (point < 10) {
    return true;
  }

  return false;
}
