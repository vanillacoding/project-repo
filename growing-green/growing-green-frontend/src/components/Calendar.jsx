import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { getFormatedDate } from '../utils/getFormatedDate';

export default function Calendar() {
  const [date, setDate] = useState({
    year: null,
    month: null,
    week: null,
    day: null,
  });

  useEffect(() => setDate(getFormatedDate(new Date())), []);

  return (
    <Wrapper>
      <div className="left-hook"></div>
      <div className="right-hook"></div>
      <div className="month-wrapper">
        <h3 className="month-name">{date.month}</h3>
      </div>
      <div className="date-wrapper">
        <h1 className="day">{date.day}</h1>
        <h3 className="week-name">{date.week}</h3>
      </div>
    </Wrapper>
  );
}

const Wrapper = styled.div`
  position: relative;
  width: 100px;
  height: 100px;
  font-family: 'Gill Sans';
  color: #2a1618;
  margin: 9rem 0 2rem 1rem;

  .left-hook {
    position: absolute;
    width: 8px;
    height: 24px;
    background: #2a1618;
    border-radius: 10px;
    top: -11px;
    left: 15px;
  }

  .right-hook {
    position: absolute;
    width: 8px;
    height: 24px;
    background: #2a1618;
    border-radius: 5px;
    top: -11px;
    right: 15px;
  }

  .month-wrapper {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 30%;
    background: ${({ theme }) => theme.baseTheme.colors.red};
    color: white;
    border-top-left-radius: 10px;
    border-top-right-radius: 10px;
    box-shadow: rgba(0, 0, 0, 0.4) 0px 2px 4px,
      rgba(0, 0, 0, 0.3) 0px 7px 13px -3px, rgba(0, 0, 0, 0.3) 3px 0px 0px inset;

    .month-name {
      margin: 0;
    }
  }

  .date-wrapper {
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    width: 100%;
    height: 70%;
    background: white;
    border-bottom-left-radius: 10px;
    border-bottom-right-radius: 10px;
    box-shadow: rgba(0, 0, 0, 0.4) 0px 2px 2px,
      rgba(0, 0, 0, 0.3) 0px 7px 3px -3px, rgba(0, 0, 0, 0.2) 3px -3px 0px inset;

    .day {
      margin: 0;
      vertical-align: middle;
      display: table-cell;
      font-size: 2.3em;
    }

    .week-name {
      font-weight: 400;
      margin-top: -0.1rem;
      margin-bottom: 0.8rem;
      font-size: 1em;
    }
  }
`;
