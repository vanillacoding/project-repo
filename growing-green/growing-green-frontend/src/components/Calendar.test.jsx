import React from 'react';
import Calendar from './Calendar';
import { render } from '../test/test-utils';
import { getFormatedDate } from '../utils/getFormatedDate';

describe('<Calendar>', () => {
  it('matches snapshot', () => {
    const { container } = render(<Calendar />);
    expect(container).toMatchSnapshot();
  });

  it('should render currnet Date', () => {
    const { getByText } = render(<Calendar />);
    const date = getFormatedDate(new Date());
    expect(getByText(date.month)).toHaveClass('month-name');
    expect(getByText(date.week)).toHaveClass('week-name');
    expect(getByText(date.day)).toHaveClass('day');
  });
});
