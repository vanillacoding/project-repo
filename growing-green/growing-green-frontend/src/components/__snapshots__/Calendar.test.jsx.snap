// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar> matches snapshot 1`] = `
<div>
  <div
    class="sc-bdvvtL fiLqdH"
  >
    <div
      class="left-hook"
    />
    <div
      class="right-hook"
    />
    <div
      class="month-wrapper"
    >
      <h3
        class="month-name"
      >
        Nov
      </h3>
    </div>
    <div
      class="date-wrapper"
    >
      <h1
        class="day"
      >
        24
      </h1>
      <h3
        class="week-name"
      >
        Wednesday
      </h3>
    </div>
  </div>
</div>
`;
